keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,285 @@
+1. `finalizers`: COSI controller adds the finalizer to defer Bucket deletion until backend deletion ops succeed.
+1. `bucketClassName`: Name of the associated BucketClass
+1. `supportedProtocols`:  String array of protocols (e.g. s3, gcs, swift, etc.) supported by the associated object store.
+1. `releasePolicy`: the release policy defined in the associated BucketClass. (see [BucketClass](#BucketClass) for more information)
+1. `bucketRef`: the name & namespace of the bound Bucket.
+1. `secretRef`: the name of the sidecar-generated secret. It's namespace is assumed `cosi-system`.
+1. `accessMode`: The level of access granted to the credentials stored in `secretRef`, one of "read only" or "read/write".
+1. `bucketAttributes`: stateful data relevant to the managing of the bucket but potentially inappropriate user knowledge (e.g. user's IAM role name)
+1. `phase`: is the current state of the BucketContent:
+    - `Bound`: the operator finished processing the request and bound the Bucket and BucketContent
+    - `Released`: the Bucket has been deleted, leaving the BucketContent unclaimed.
+    - `Failed`: error and all retries have been exhausted.
+    - `Retrying`: set when a recoverable driver or kubernetes error is encountered during bucket creation or access granting. Will be retried.
+
+> \* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+
+#### BucketClass
+
+A cluster-scoped custom resource. The BucketClass defines a release policy, and specifies driver specific parameters, such as region, bucket lifecycle policies, etc., as well as the name of the driver.  The driver name is used to filter BuckeContenst meant to be handled by a given driver.  In static bucket workflows, the driver name may be empty if the object bucket is defined.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketClass
+metadata:
+  name: 
+provisioner: [1]
+supportedProtocols: [2]
+accessMode: {"ro", "rw"} [3]
+releasePolicy: {"Delete", "Retain"} [4]
+bucketContentName: [5]
+parameters: string:string [6]
+```
+
+1. `provisioner`: Used by sidecars to filter BucketContent objects 
+1. `supportedProtocols`: A strings array of protocols the associated object store supports (e.g. swift, s3, gcs, etc.)
+1. `accessModes`: Declares the level of access given to credentials provisioned through this class.
+1.  `releasePolicy`: Prescribes outcome of a Deletion and Revoke events.
+    - `Delete`:  the bucket and its contents are destroyed
+    - `Retain`:  the bucket and its contents are preserved, only the user’s access privileges are terminated
+- `bucketContentName`: (Optional). An admin defined BucketContent representing a Brownfield or Static bucket.  A non-nil value in this field prevents the BucketClass from being used for Greenfield.
+- `parameters`: object store specific key-value pairs passed to the driver.
```

Comment:
This is where a secret name + ns would be defined for object store using a secret to contain the cred needed to create buckets.

Reply:
IMO the parameter list is exclusively used to store driver defined variables.  A discrete field should be included for a secreteRef.

Related - what does this secret represent?  Is it an administrative credential used to affect user policy and bucket creation?  Or is it a user privileged cred meant to be associated with the bucket?  Particularly in brownfield/static, is it already associated with the bucket as a user cred?